codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@nmittler - Thanks for clarifying. I knew what you meant by "protecting" the queue but I just wanted to emphasize the balance between preserving an existing interface + working around issues, or just changing the interface to adapt to event flow (should have phrased it this way originally...sorry :) ).  In this situation since the interface (meaning the methods and the contract is operates by) is under our control and I think changing it is an option. It doesn't have to be what I have suggested, but if changing the interface can keep things simple (reduce state, protecting from the outside, etc...) lets explore this option.

I understand the "fragile" point you made.  I am focused on how we will reduce the fragility (and ease of implementation of the interface) by allowing these methods to be called and just following the error path rather than introduce extra state, and code paths to be able to compensate as we are now.

> Maybe we can somehow be aware that we're in the middle of a write callback and postpone clearing the queue until after it completes?

I'm not opposed to this suggestion but I just wonder how much additional state and sequencing dependencies we will need to add.  For example tracking when we are in a write, maybe tracking an exception that is raised if a clear happens, then checking some state after the write if we need to take any action?  Does this approach complicate things if other (write future) exceptions occur before/after this?